File: golang/count_binary_substrings.go
package golang

// Space : O(1)
// Time : O(n)

func countBinarySubstrings(s string) int {
	ans := 0
	zero, one := 0, 0
	n := len(s)

	last := string(s[0])
	if last == "0" {
		zero++
	} else {
		one++
	}

	for i := 1; i < n; i++ {
		if string(s[i]) != last {
			ans += Min(one, zero)
			if string(s[i]) == "0" {
				zero = 1
			} else {
				one = 1
			}
		} else {
			if string(s[i]) == "0" {
				zero++
			} else {
				one++
			}
		}
		last = string(s[i])
	}

	ans += Min(one, zero)
	return ans
}

File: python3/count_binary_substrings.py
# Space : O(1)
# Time : O(n)

class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        ans = 0
        last = s[0]
        one = 1 if s[0] == '1' else 0
        zero = 1 if s[0] == '0' else 0
        n = len(s)
        
        for i in range(1, n):
            if s[i] != last:
                ans += min(one, zero)
                if s[i] == '0':
                    zero = 1
                else:
                    one = 1
            else:
                if s[i] == '0':
                    zero += 1
                else:
                    one += 1
            last = s[i]

        ans += min(one, zero)
        
        return ans